lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2459,30 +2459,156 @@ NamingPatternRequest::evaluate(Evaluator &evaluator, VarDecl *VD) const {
 
 namespace {
 
-// Utility class for deterministically ordering vtable entries for
-// synthesized methods.
-struct SortedFuncList {
-  using Key = std::tuple<DeclName, std::string>;
-  using Entry = std::pair<Key, AbstractFunctionDecl *>;
+/// A kind that's used to roughly sort the members of
+enum class SortedDeclKind: uint8_t {
+  EnumCase,
+  Type,
+  Initializer,
+  Deinitializer,
+  Property,
+  Function,
+  Operator,
+  Other
+};
+
+// Utility class for deterministically ordering implicit entries.
+struct SortedDeclList {
+  using Entry = std::pair<ValueDecl *, std::string>;
   SmallVector<Entry, 2> elts;
   bool sorted = false;
 
-  void add(AbstractFunctionDecl *afd) {
-    assert(!isa<AccessorDecl>(afd));
-
-    Key key{afd->getName(), afd->getInterfaceType().getString()};
-    elts.emplace_back(key, afd);
+  void add(ValueDecl *value) {
+    assert(!isa<AccessorDecl>(value));
+    assert(std::find_if(elts.begin(), elts.end(), [&](const Entry &entry) {
+      return entry.first == value;
+    }) == elts.end());
+    elts.push_back({value, std::string()});
   }
 
   bool empty() { return elts.empty(); }
 
+  static SortedDeclKind getSortedDeclKind(ValueDecl *valueDecl){
+    switch (valueDecl->getKind()) {
+      case DeclKind::EnumCase:
+      case DeclKind::EnumElement:
+        return SortedDeclKind::EnumCase;
+
+      case DeclKind::AssociatedType:
+      case DeclKind::Class:
+      case DeclKind::Enum:
+      case DeclKind::GenericTypeParam:
+      case DeclKind::OpaqueType:
+      case DeclKind::Protocol:
+      case DeclKind::Struct:
+      case DeclKind::TypeAlias:
+        return SortedDeclKind::Type;
+
+      case DeclKind::Constructor:
+        return SortedDeclKind::Initializer;
+
+      case DeclKind::Destructor:
+        return SortedDeclKind::Deinitializer;
+
+      case DeclKind::Param:
+      case DeclKind::PatternBinding:
+      case DeclKind::Var:
+        return SortedDeclKind::Property;
+
+      case DeclKind::Accessor:
+        return SortedDeclKind::Function;
+
+      case DeclKind::Func:
+        return cast<FuncDecl>(valueDecl)->isOperator()
+            ? SortedDeclKind::Operator
+            : SortedDeclKind::Function;
+
+
+      case DeclKind::InfixOperator:
+      case DeclKind::PrecedenceGroup:
+      case DeclKind::PrefixOperator:
+      case DeclKind::PostfixOperator:
+      case DeclKind::Subscript:
+        return SortedDeclKind::Operator;
+
+      case DeclKind::Extension:
+      case DeclKind::IfConfig:
+      case DeclKind::Import:
+      case DeclKind::MissingMember:
+      case DeclKind::Module:
+      case DeclKind::TopLevelCode:
+      case DeclKind::PoundDiagnostic:
+        return SortedDeclKind::Other;
+    }
+  }
```

Comment:
It looks like the only time one `DeclKind` -> two `SortedDeclKinds` is for `Func`. Would it be possible to just split the `DeclKind::Func` into `DeclKind:NamedFunc` and `DeclKind::Operator`? We could keep `DeclFunc` as an abstract class. Then, maybe we wouldn't need `SortedDeclKind` at all.

Reply:
That refactor isn't interesting to any other clients, and would not eliminate `SortedDeclKind` because we still want the ordering to be defined even if `DeclKind` gets refactored.